orttraining/orttraining/core/session/training_session.cc

```diff
@@ -130,7 +130,18 @@ Status TrainingSession::ConfigureForTraining(
                                          config.distributed_config.horizontal_parallel_size,
                                          config.distributed_config.pipeline_parallel_size});
 
-  ORT_RETURN_IF_ERROR(ApplyTransformationsToMainGraph());
+  // We need to get trainable weights to prevent constant folding from them. This works well if trainable weights are passed from config.
+  // For case we use GetTrainableModelInitializers to get trainable weights such as C++ frontend, it may get more initializers
+  // than trainable weights here as it's before transformers. So the constant folding may miss some nodes we actually can fold.
```

Comment:
abpout 
about 

> For case we use GetTrainableModelInitializers to get trainable weights such as C++ frontend, it may get more initializers
>  than trainable weights here as it's before transformers

I think the real reason  is: the immutable config in c++ frontend is targeting on transformed graph, this is not a normal practice I think. Ideally if we pass the immutable nodes target original graph, but one restriction would be we cannot cover bw graph. So get back further, confgi.gimmutable_weights might better to replaced by weight_names_to_train, for c++ frontend as I suggested offline. Though I am fine with this workaround, we can fix that later if necessary.  @SherlockNoMad